### ISSUE TITLE
Refactoring. Make GenerateAllCombinations method more developer friendly

### ISSUE THREAD

**Issue by skoshelev** [nopSolutions/nopCommerce#2932]
At the moment, this method is very difficult to understand, it contains bitwise operations and generates unnecessary data.

**Comment by Meakeel** [comment]
Hello!

When refactoring could you look at adding a better version of this to the function before all the combinations are generated.

When you have a product that uses a high number of attributes (26 in this case each with 2-10 values each) then the whole system fails very quickly (Memory overflow)

This cuts down the generate to only use the attrbiutes that have been selected.

```
            if (allowedAttributeIds?.Any() ?? false)
            {
                allProductAttributMappings = allProductAttributMappings.Where(x => allowedAttributeIds.Any(t => x.ProductAttributeValues.Any(c => c.Id.Equals(t)))).ToList();
            }```

**Comment by skoshelev** [comment]
Closed #2932
